pkg/registry/service/pool_allocator.go

```diff
@@ -0,0 +1,306 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"fmt"
+	math_rand "math/rand"
+	"sync"
+
+	"os"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+	"strings"
+)
+
+// A PoolAllocator is a helper interface that supports a pool of resources that can be allocated
+type PoolAllocator interface {
+	// Allocate allocates a specific entry.
+	Allocate(s string) error
+
+	// AllocateNext allocates and returns a new entry.
+	AllocateNext() (string, error)
+
+	// Release de-allocates an entry.
+	Release(s string) (bool, error)
+}
+
+// A very simple string iterator
+type StringIterator interface {
+	Next() (string, bool)
+}
+
+// A PoolDriver provides the PoolAllocator with the set from which it allocates
+type PoolDriver interface {
+	// Choose an item from the set
+	PickRandom(random *math_rand.Rand) string
+
+	// Iterate across all items in the set
+	Iterate() StringIterator
+}
+
+// MemoryPoolAllocator is PoolAllocator that is backed by an in-memory collection,
+// can't be used if there are multiple processes allocating from the pool (cf EtcdPoolAllocator)
+type MemoryPoolAllocator struct {
+	driver PoolDriver
+
+	lock sync.Mutex // protects 'used' and 'random'
+
+	used           map[string]bool
+	randomAttempts int
+
+	random *math_rand.Rand
+}
+
+// EtcdPoolAllocator is PoolAllocator that is backed by etcd, so can be used by multiple processes
+type EtcdPoolAllocator struct {
```

Comment:
Etcd has to be in a separate package - we keep it isolated to make future refactoring easier.  So back to `allocator/etcd`